psi4/src/psi4/libfmm/multipoles_helper.h

```diff
@@ -0,0 +1,217 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Psi4: an open-source quantum chemistry software package
+ *
+ * Copyright (c) 2007-2022 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#ifndef libfmm_mpoles_helper_H
+#define libfmm_mpoles_helper_H
+
+#include "psi4/pragma.h"
+
+#include "psi4/libmints/vector.h"
+#include "psi4/libmints/vector3.h"
+#include "psi4/libmints/matrix.h"
+
+#include <functional>
+#include <memory>
+#include <tuple>
+#include <vector>
+#include <unordered_map>
+
+namespace psi {
+
+enum SolidHarmonicsType {Regular, Irregular};
+
+// Copied stuff from libmints/solidharmonics.cc
+static inline int npure(int l) { return 2 * l + 1; }
+static inline int icart(int a, int b, int c) { return (((((a + b + c + 1) << 1) - a) * (a + 1)) >> 1) - b - 1; }
+static inline int ipure(int, int m) { return m < 0 ? 2 * -m : (m == 0 ? 0 : 2 * m - 1); }
+
+// Some more helper functions
+static inline int ncart(int l) { return (l+1)*(l+2)/2; }
+
+// Some more useful Helper Functions
+static int choose(int n, int r) {
+    if (r < 0 || r > n) {
+        return 0;
+    }
+    int small = std::min(n, n-r);
+    int nCr = 1;
+    for (int t = 0; t < small; t++) {
+        nCr *= n;
+        nCr /= (t+1);
+        n -= 1;
+    }
+    return nCr;
+}
+
+static int m_addr(int m) {
+    /*- Return the unsigned (array) address of m -*/
+    if (m <= 0) {
+        // 0, 1s, 2s, 3s, ...
+        return 2*(-m);
+    } else {
+        // 1c, 2c, 3c, ...
+        return 2*m-1;
+    }
+}
+
+extern double factorial(int n);
+
+class PSI_API MultipoleRotationFactory {
+
+    protected:
+      Vector3 R_a_;
+      Vector3 R_b_;
+
+      // New Z axis in rotated frame of reference
+      SharedMatrix Uz_;
+
+      // Maximal Angular Momentum
+      int lmax_;
+
+      // Cached Rotation Matrices in a vector of Matrices
+      std::vector<SharedMatrix> D_cache_;
+
+      double U(int l, int m, int M);
+      double V(int l, int m, int M);
+      double W(int l, int m, int M);
+      double P(int i, int l, int mu, int M);
```

Comment:
Could we get either a very brief explanation of what these functions do, or a reference to the paper/book in which they're defined (ideally w/ equation numbers)?

Reply:
Done